### DISCUSSION TITLE
Request for maintainers: word import open source logic

### DISCUSSION THREAD

**Discussion by TheSpyder** [tinymce/tinymce#7487] [Announcements]
Hi all,

As announced two months ago we are removing the word import logic from our open source paste plugin in TinyMCE 6:
https://www.tiny.cloud/blog/rich-text-editor-open-source-5-6-0-release

There are business reasons for this, but I'm here to discuss the technical reasons and propose a path forward. To give some context to this decision, we decided to merge the `paste` plugin into the editor core with TinyMCE 6 to gain a number of benefits:

* Paste is a critical feature for any rich text editor, without the `paste` plugin TinyMCE's paste experience is... not great.
* Because paste is a plugin it must operate in a way that is incompatible with other paste plugins, to the point where we had to [add code to the paste plugin that detects our premium offering and disables itself](https://github.com/tinymce/tinymce/blob/9df1e5617676ac10d49ebff7e5fff7a5484818ff/modules/tinymce/src/plugins/paste/main/ts/alien/DetectProPlugin.ts). This will no longer be a problem when paste is a core feature.
* As a result of needing to disable `paste` when using an external paste plugin, any external paste offering (including our own) was required to replicate the baseline `paste` plugin behaviour on top of any new behaviour it provides. With this change, that will no longer be necessary.

When we were planning this change we realised it would be difficult to have word filtering in the core while also making sure it could be overridden in a plugin. It would also greatly increase the bundle size impact of merging `paste` into the core. This seemed like a good opportunity to apply the business reasoning and delete it.

The work to remove the word import logic has already been done on the v6 branch (#7371) and we are currently implementing the change that combines paste into core.

While in our opinion there are mostly positive benefits to this change we understand the decision will upset users relying on our open source word filtering. We would like to offer the community an opportunity to take over maintenance of this logic. The deal would be:

* take the [TinyMCE 5 version of the paste plugin](https://github.com/tinymce/tinymce/tree/5.10.2/modules/tinymce/src/plugins/paste)
* separate out the [Word Filtering logic](https://github.com/tinymce/tinymce/tree/5.10.2/modules/tinymce/src/plugins/paste/main/ts/core/WordFilter.ts) and any other parts of the plugin required for standalone filtering
* create a new plugin and repository based on it
* the TinyMCE 6 migration guide will link to the new community offering alongside our premium offering as options for developers upgrading to TinyMCE 6
* we are not offering this code under a new license, so the new project will need to comply with the LGPL

I am happy to answer any questions about this, publicly or privately (andrew.herron@tiny.cloud).

Regards,
Andy

---
Andrew Herron
Staff Software Engineer

**Comment by t00** [comment]
Is there a way to keep a stock TinyMCS 6 paste plugin and just intercept paste action to manually filter out pasted document, using a TinyMCE 5 code from the mentioned repository?

**Reply by TheSpyder** [reply]
Yes. The paste plugin was integrated into core with TinyMCE 6. As I described, it should be possible to reduce the TinyMCE 5 paste plugin to just the MS Word filtering logic and build a plugin on that (this is similar to how our premium PowerPaste plugin operates). Something along the lines of:
```
  editor.on('paste', (e) => {
    if (pastingFromMSWord(e)) {
      handlePaste(e);
      e.preventDefault();
    }
  });
```

**Comment by lubomirblazekcz** [comment]
There is an open-source plugin now https://github.com/openregion/tinymce-word-paste-plugin, maybe the fixes by @aautio from https://github.com/tinymce/tinymce/pull/6690 can be implemented there?

**Reply by TheSpyder** [reply]
Oh that's good to know! They didn't reach out to us, as far as I know, to be nominated as the community replacement for the paste plugin. But I would be interested to hear how well it works.

**Reply by okainov** [reply]
@TheSpyder even though you haven't heard from the maintainer of this plugin, it seems to do its job just great, so I advocate for mentioning it in the documentation page!

**Reply by TheSpyder** [reply]
I'm not going to send that sort of message to our users unless the maintainer asks for it. Seems a bit rude.
